File: go/vertex_cover.go
package main

import (
    "fmt"
    "container/list"
)

func main(){

    fmt.Printf("Parameterized Vertex Cover\n")
    fmt.Printf("Enter n:")
    var n int
    fmt.Scanf("%d\n", &n)

    //make Graph using map.
    var graph map[int]*list.List
    graph = make(map[int]*list.List)

    fmt.Printf("Enter edges:\n")

    for i := 0;  i < n; i++ {
        var a, b int
        fmt.Scanf("%d %d\n", &a, &b)
        //fmt.Printf("%d %d\n", a, b)
        l , ok := graph[a]
        if ok {
            l.PushBack(b)
        }else{
            graph[a] = list.New()
            graph[a].PushBack(b)
        }

        l, ok = graph[b]
        if ok {
            l.PushBack(a)
        }else{
            graph[b] = list.New()
            graph[b].PushBack(a)
        }
    }

    fmt.Printf("Enter the parameter k: ")
    var k int
    fmt.Scanf("%d\n", &k)

    vertex_cover(graph,k)



}

// depth bounded search
func vertex_cover(graph map[int]*list.List, k int){

    //reduction one  : Remove degree zero vertices
    count := 0
    for k,v := range graph {
        if v.Len() == 0 {
            fmt.Printf("Delete %d\n",k)
            delete(graph, k)
            count++
        }
    }

    //reduction two : Remove degree k+1 or more vertices

    for node,neighbours := range graph {
        if neighbours.Len() >= (k+1) {

            fmt.Printf("Delete %d\n",node)

            for e := neighbours.Front(); e != nil; e = e.Next() {

                v := e.Value.(int) // need to master casting
                //graph[v].Remove(node)
                fmt.Println(v)
            }
            delete(graph, node)
            count++
        }
    }




    fmt.Println(graph)

}

File: java/src/jav/VertexCover.java
package jav;


import java.util.*;
import java.util.Map.Entry;


public class VertexCover {
	
	
	public static void main(String[] args) {
		
		Scanner sc = new Scanner(System.in);
		System.out.print("Enter n:");
		int n = sc.nextInt();
		Graph graph = new Graph();
		
		System.out.println("Enter edges: ");
		for(int i = 0;i <n;i++){
			int a = sc.nextInt();
			int b = sc.nextInt();
			graph.addEdge(a,b);
		}
		
		
	}
	
	public static boolean vertex_cover(Graph graph, int k){
		try{
			
			if(k<0) return false;
			
			//redution one and two
			//remove degree zero vertices and those with degree > k
			int k_dash = k;
			Iterator<Entry<Integer, HashSet<Integer>>> iter = graph.iterator();
			
			while(iter.hasNext()){
				Entry<Integer, HashSet<Integer>> entry = iter.next();
				int node = entry.getKey();
				HashSet<Integer> neighbours = entry.getValue();
				if(neighbours.size()==0){
					System.out.println("remove"+node);
					graph.remove(node, neighbours, false);
					iter.remove();
				}
			}
			
			iter = graph.iterator();
			
			while(iter.hasNext()){
				Entry<Integer, HashSet<Integer>> entry = iter.next();
				int node = entry.getKey();
				HashSet<Integer> neighbours = entry.getValue();
				if(neighbours.size()>k){
					System.out.println("remove"+node);
					graph.remove(node, neighbours, false);
					iter.remove();
					k_dash--;
				}
			}
			
			if(k_dash<0||graph.edges.size()>k_dash*k_dash) return false;
				
			if(graph.edges.size()>0){
				System.out.println(graph.edges.size());
				Iterator<Edge> it = graph.edges.iterator();
				if(!it.hasNext()){
					System.out.println("Should be the case.");
					return false;
				}
				Edge e = it.next();
				System.out.println("e:"+e.u+":"+e.v);
					Graph g1 = (Graph) graph.clone();
					
					g1.remove(e.u,g1.graph.get(e.u),true);
					if(vertex_cover(g1, k_dash)){
						return true;
					}
	
					Graph g2 = (Graph) graph.clone();
					g2.remove(e.v, g2.graph.get(e.v), true);
					return  vertex_cover(g2, k_dash);
					
				
							
			}else{
				//no edge to cover
				System.out.println(graph.edges);
				return true;
			}
		
	  }catch(Exception e){
		  e.printStackTrace();
		  return false;
	  }

	}

}
